File: Core/src/main/java/jokatu/game/GameID.java
package jokatu.game;

import com.fasterxml.jackson.annotation.JsonValue;
import jokatu.identity.Identity;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * An object that uniquely identifies a game.
 * @author Steven Weston
 */
public class GameID implements Identity, Comparable<GameID> {

	private long identity;

	@SuppressWarnings("unused") // This is called by Jackson
	public GameID() {}

	public GameID(long identity) {setIdentity(identity);}

	@SuppressWarnings("unused") // This is called by Jackson
	@JsonValue
	public long getIdentity() {
		return identity;
	}

	public void setIdentity(long identity) {
		this.identity = identity;
	}

	@Override
	public boolean equals(@Nullable Object o) {

		if (this == o) {
			return true;
		}
		if (o == null || getClass() != o.getClass()) {
			return false;
		}

		GameID gameID = (GameID) o;

		return identity == gameID.identity;
	}

	@Override
	public int hashCode() {
		return (int) (identity ^ (identity >>> 32));
	}

	@Override
	public String toString() {
		return Long.toString(identity);
	}

	@Override
	public int compareTo(@NotNull GameID o) {
		if (identity < o.identity) {
			return -1;
		} else if (identity > o.identity) {
			return 1;
		} else {
			return 0;
		}
	}
}

File: Core/src/main/kotlin/jokatu/game/GameID.kt
package jokatu.game

import com.fasterxml.jackson.annotation.JsonValue
import jokatu.identity.Identity
import java.lang.Long.toString

/**
 * An object that uniquely identifies a game.
 * @author Steven Weston
 */
class GameID : Identity, Comparable<GameID> {

	// This is called by Jackson
	@get:JsonValue
	var identity: Long = 0

	// This is called by Jackson
	constructor()

	constructor(identity: Long) {
		this.identity = identity
	}

	override fun equals(other: Any?): Boolean {

		if (this === other) {
			return true
		}
		if (other == null || javaClass != other.javaClass) {
			return false
		}

		val gameID = other as GameID?

		return identity == gameID!!.identity
	}

	override fun hashCode(): Int {
		return (identity xor identity.ushr(32)).toInt()
	}

	override fun toString(): String {
		return toString(identity)
	}

	override fun compareTo(other: GameID): Int {
		return when {
			identity < other.identity -> -1
			identity > other.identity -> 1
			else -> 0
		}
	}
}
